doc/rfc001-sum-types.md

```diff
@@ -0,0 +1,846 @@
+def describeInt(n: int): str = if (n < 0) then "negative" else "positive"
+val ex : str = match e {
+  a : x => describeInt(x),
+  b : x => x,
+}
+```
+
+A textbook rule for eliminating an expression that is a variant of a finite sum
+type can be given as
+
+$$ 
+\frac{
+\Gamma \vdash e \colon 
+\langle i_1 \hookrightarrow \tau_1, \ldots, i_n \hookrightarrow \tau_n \rangle 
+\quad 
+\Gamma, x_1 : \tau_1 \vdash e_1 \colon \tau
+\quad 
+\ldots
+\quad 
+\Gamma, x_n : \tau_n \vdash e_n \colon \tau
+}
+{ \Gamma \vdash \ match \ e \ 
+\\{ i_1 \cdot x_1 \hookrightarrow e_1 | \ldots | i_n \cdot x_n \hookrightarrow e_n \\} : \tau }
+$$
+
+Note the complementary symmetry compared with the textbook rule for product
+construction: product construction requires `n` expressions to conclude with a
+single record-type expression combining them all; while sum type elimination
+requires a single sum-typed expression and `n` ways to convert each of the `n`
+alternatives of the sum type to conclude with a single expression of a type that
+does not need to appear in the sum type at all.
+
+The proposed rule for quint's type system is given without an attempt to
+reproduce our practice of using quint strings. This can be added in later if needed:
+
+$$
+\frac{
+\Gamma \vdash e : (s, c) \quad
+\Gamma, x_1 \vdash e_1 : (t, c_1) \quad \ldots \quad \Gamma, x_n \vdash e_n : (t, c_n) \quad
+\Gamma, \langle v \rangle \vdash e_{n+1} : (t, c_{n+1}) \quad
+fresh(v)
+}{
+\Gamma \vdash \ match \ e \ \\{ i_1 : x_1 \Rightarrow e_1, \ldots, i_n : x_n \Rightarrow e_n \\} : (t, 
+c \land c_1 \land \ldots \land c_n \land c_{n+1} \land
+s \sim \langle i_1 : t_1, \ldots, i_n : t_n | v \rangle)
+}
+$$
+
+Compared with quint's rule for product construction we see the same
+complementary symmetry. However, we also see a striking difference: there is no
+row variable occurring in the product construction, but the row variable plays
+an essential function in sum type elimination of row-based variants. Row types
+in records are useful for extension operations (i.e., which we don't support in
+quint currently) and for operators that work over some known fields but leave
+the rest of the record contents variable. But the core idea formalized in a
+product type is that the constructor _must_ package all the specified things
+together so that the recipient _can_ chose any thing; thus, when a record is
+constructed we must supply all the things and there is no room for variability
+in the row. For sum types, in contrast the constructor _can_ supply any one thing (of
+a valid alternate type), and requires the recipient _must_ be prepared to
+handle every possible alternative. 
+
+In the presence of row-polymorphis, however, the responsibility of the recipient
+is relaxed: the recipient can just handle a subset of the possible alternatives,
+and if the expression falls under a label they are not prepared to handle, they
+can pass the remaining responsibility on to another handler.
```

Comment:
This explanation makes sense to me. However, I am not sure how we would handle that in Apalache. Maybe it is not a problem at all, since Apalache variants would require us to write series of `IF-THEN-ELSE` with unsafe access, but we should check that.

Reply:
We can translate this into Apalache IR using the `VariantGetUnsafe` and `VariantTag` operators, like the pattern here:

https://github.com/informalsystems/apalache/blob/4e0094ee5d03a068188a85dcdc8ead021b17033e/src/tla/Option.tla#L34-L60